packages/block-library/src/cover/edit/index.js

```diff
@@ -296,8 +389,8 @@ function CoverEdit( {
 			backgroundType: useFeaturedImage
 				? IMAGE_BACKGROUND_TYPE
 				: undefined,
-			isDark: isDarkSetting,
 		} );
+		setIsDark( overlayColor.color, await getAverageBackgroundColor( url ) );
```

Comment:
Moving the await below the other `setAttributes` is going to cause the attributes to be update in two steps with a delay vs one step before. I haven't run the code yet, but I expect this will trigger two renders instead of one.

The effect of everything else being updated and then the `isDark` setting lagging behind will be more obvious on slower hardware. Moving the `await` to the top will probably have the same behavior as before, but I'm not 100% certain because `setAttributes` is weird and it would still get called twice, just closer together.

Same thing for `onSetOverlayColor` and `onUpdateDimRatio`.

Reply:
also good to think about making atomic updates so that state is never partially updated and thus inconsistent. we definitely only want one undo/redo point for this operation, if any at all (which I guess we might not want an undo step since this flows from the other user-initiated action?)

Reply:
I updated some places, but I had to draw the line somewhere with all the refactoring that was required. And the line I drew was refactoring out `withColors` which is a HOC that adds the `setOverlayColor` prop which calls `setAttributes`.

I made sure that everywhere that `setOverlayColor` was followed by another `setAttributes`, there was an `__unstableMarkNextChangeAsNotPersistent()` call to keep the undo levels to what you would expect while editing.